scylla-cdc/src/stream_generations.rs

```diff
@@ -0,0 +1,531 @@
+use futures::stream::StreamExt;
+use scylla::batch::Consistency;
+use scylla::frame::response::result::Row;
+use scylla::frame::value::Timestamp;
+use scylla::query::Query;
+use scylla::{IntoTypedRows, Session};
+use std::error::Error;
+use std::sync::Arc;
+use uuid::Uuid;
+
+#[derive(Debug, Clone, Eq, PartialEq, Hash, Ord, PartialOrd)]
+pub struct Generation(chrono::Duration);
+
+#[derive(Debug, Clone, Eq, PartialEq, Hash, Ord, PartialOrd)]
+pub struct StreamID(Vec<u8>);
+
+// Component responsible for managing stream generations.
+pub struct GenerationFetcher {
+    generations_table_name: String,
+    streams_table_name: String,
+    session: Arc<Session>,
+}
+
+// Number taken from: https://www.scylladb.com/2017/11/17/7-rules-planning-queries-maximum-performance/.
+const DEFAULT_PAGE_SIZE: i32 = 5000;
+
+impl GenerationFetcher {
+    // Creates new instance of GenerationFetcher.
+    pub fn new(session: &Arc<Session>) -> GenerationFetcher {
+        GenerationFetcher {
+            generations_table_name: "system_distributed.cdc_generation_timestamps".to_string(),
+            streams_table_name: "system_distributed.cdc_streams_descriptions_v2".to_string(),
+            session: Arc::clone(session),
+        }
+    }
+
+    // Function instead of constant for testing purposes.
+    fn get_all_stream_generations_query(&self) -> String {
+        format!(
+            r#"
+    SELECT time
+    FROM {}
+    WHERE key = 'timestamps';
+    "#,
+            self.generations_table_name
+        )
+    }
+
+    pub async fn fetch_all_generations(&self) -> Result<Vec<Generation>, Box<dyn Error>> {
+        let mut query = Query::new(self.get_all_stream_generations_query());
+        self.select_consistency(&mut query).await?;
+        query.set_page_size(DEFAULT_PAGE_SIZE);
+
+        let mut rows = self
+            .session
+            .query_iter(query, &[])
+            .await?
+            .into_typed::<(chrono::Duration,)>();
+
+        let mut generations = vec![];
+
+        while let Some(next_row) = rows.next().await {
+            let (generation,) = next_row?;
+            generations.push(Generation(generation));
+        }
+
+        Ok(generations)
+    }
+
+    // Function instead of constant for testing purposes.
+    fn get_generation_by_timestamp_query(&self) -> String {
+        format!(
+            r#"
+    SELECT time
+    FROM {}
+    WHERE key = 'timestamps'
+    AND time <= ?
```

Comment:
Please add an explicit `ORDER BY time DESC` here. When creating a new table without specifying the ordering of clustering key, the clustering key is ordered ASC by default - the generations table is ordered DESC, so an explicit mention of ordering should reduce confusion.